## Defer

[[Defer] https://go.dev/tour/flowcontrol/12](https://go.dev/tour/flowcontrol/12)

[[Stacking defers] https://go.dev/tour/flowcontrol/13](https://go.dev/tour/flowcontrol/13)

> The deferred call's arguments are evaluated immediately, but the function call is not executed until the surrounding function returns.

The page includes the note above but the example on the page does not reflect the essence of this warning. A more appropriate example would be like the following.
```go
package main

import "fmt"

func main() {
	message := "world"
	defer fmt.Println(message)
	message = "hello"
	fmt.Println(message)
}
```
[Run on Go Playground](https://go.dev/play/p/Eid3X3-iCqM)
```
hello
world

Program exited.
```

As you can see in the output, although the `defer fmt.Println(message)` statement is executed just before leaving the `main()` function's scope, the value of the variable `message` evaluated where the `defer` statement actually sits in the code.

If defer just postpones the execution of the statement why use defer instead of putting it just before the curly bracket? Having multiple return statements within a function is idiomatic. So there's usually no guarantee that the flow of the product's execution reach the line just before the curly bracket. Think of a function where you query your database to find a record. The best case flow of this function would be like this:
- Open a connection to the database
- Make your query
- Close the database connection
- Return the record.

But lots of things may go wrong with a database connection. You'll want to catch any errors returning from the query method and return prematurely. But you'll always want to close the connection to prevent spilling. Instead of adding the connection closing code to each and every possible return from the function, it's always better to defer once just after the opening of connection and stay safe. So a better flow would look like this:
- Open a connection to the database
- Check for error
  - If there's an error, return error
  - Else, defer closing connection
- Make your query
  - If there's and error return error
  - Else, do whatever you want to do with the record and return

Defer is a lifesaver, if you use it of course. This means that you have to be aware of any resource that needs to be cleaned up upon exit. The following code snippet queries a mongodb database. The `Find` function returns a `cursor` that holds one or more records. Inspecting the go package for the `cursor` the reveals that it has a `Close` function, also points out that `Next` and `TryNext` functions must not be called after the `Close` function, which makes `Close` function a good candidate for `defer`.
[https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.11.3/mongo#Cursor.Close](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.11.3/mongo#Cursor.Close)
```go
    cursor, err := collection.Find(ctx, bson.M{}, opts)
    if err != nil {
        log.Error().Err(err).Msgf("error getting Records")
        return nil, err
    }
    records := make([]Record, 0)
    defer cursor.Close(ctx)
    if err = cursor.All(ctx, &records); err != nil {
        return nil, err
    }
    return records, nil
```

Unlike `Find` function, `FindOne` function returns a `SingleResult` object, which does not offer a `Close` function. 
[https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.11.3/mongo#SingleResult](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.11.3/mongo#SingleResult)

```go
    var record Record
    err = collection.FindOne(ctx, bson.M{"_id": objID}).Decode(&record)
    if err != nil {
        log.Error().Err(err).Msgf("error getting record with ID: %s", ID)
        return Record{}, err
    }
    return record, nil
```

The difference between the behaviours of those `Cursor` and `SingleResult` objects are how they are implemented. Because `Cursor` is designed to hold a number of records, which can be quite large, it is implemented in a way that it gets those records in batches from the database as it is iterated, not all at once. This is clearly stated in the MongoDB documentation on the `Cursor` object:

> A cursor is a mechanism that allows an application to iterate over database results while holding only a subset of them in memory at a given time. Read operations that match multiple documents use a cursor to return those documents in batches rather than all at once.
[https://www.mongodb.com/docs/drivers/go/current/fundamentals/crud/read-operations/cursor/](https://www.mongodb.com/docs/drivers/go/current/fundamentals/crud/read-operations/cursor/)

The key takeaways here are
* Always read the documentation before using anything.
* Always be on the lookout on those objects that might be needed closing. Especially with things like database connections, streams, filesystem resources, it's quite common to remain open until you close them explicitly.

What happens if you fail to close those objects in your code? It leads to memory leaks, filling of the connection pool of the system you're connecting to, blockage of access to a resource are the most common results. Especially if your program runs to serve multiple requests concurrently (like a web service) it will be a matter of minutes if not seconds that your application encounter a runtime error.


### Panic
> Panic is a built-in function that stops the ordinary flow of control and begins panicking.

You probably saw some panics (hopefully not so often) from the standard library. When something happens which is not recoverable, e.g. divide by zero. Also, you can use `panic` keyword to cause panic on your program. You may be wondering why I brought up panics while talking about defer. It's essential to know that deferred statements will run even if the program panics.
```go
package main

import "fmt"

func main() {
	defer fmt.Println("Write this whatever happens")
	zero := 0
	t := 10 / zero // This line will cause panic
	fmt.Println(t)
}
```
```
Write this whatever happens
panic: runtime error: integer divide by zero

goroutine 1 [running]:
main.main()
	/tmp/sandbox3713112275/prog.go:8 +0x6c

Program exited.
```
[Run on Go Playground](https://go.dev/play/p/dUgsKth9uyL)

### Context
> [Package context](https://pkg.go.dev/context) defines the Context type, which carries deadlines, cancellation signals, and other request-scoped values across API boundaries and between processes.

There are two points worth mentioning about context package while we are talking about defer keyword. First, every context creation comes with its cancel method, which should be deferred immediately even if the context has a deadline. Even the example on context package comments this.

> Even though ctx will be expired, it is good practice to call its 
> cancellation function in any case. Failure to do so may keep the 
> context and its parent alive longer than necessary.

[Package context / func WithDeadline](https://pkg.go.dev/context#WithDeadline)

Secondly, it's important to note that deferred functions will be triggered on a context expire. So using defer is again a safe way to ensure some function to run even if your function's ending is decided by a context.
```go
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	d := time.Now().Add(2 * time.Second)
	ctx, cancel := context.WithDeadline(context.Background(), d)
	defer cancel()
	fmt.Printf("Calling delay function. Time signature is %s\n", time.Now().Format("2006-01-02 15:04:05"))
	delay(ctx)
}

func delay(ctx context.Context) {
	defer fmt.Printf("This is the end of delay function. Time signature is %s\n", time.Now().Format("2006-01-02 15:04:05"))
	select {
	case <-time.After(1 * time.Second):
		fmt.Println("sleeping")
	case <-ctx.Done():
		return
	}
}
```
[Run on Go Playground](https://go.dev/play/p/xb8EY46LNyc)
```
Calling delay function. Time signature is 2009-11-10 23:00:00
sleeping
This is the end of delay function. Time signature is 2009-11-10 23:00:00
```
Even the time functions doesn't work correctly on Go Playground, we can clearly see when the context is `Done` and just before the function returns, the deferred `fmt.Printf` line is run. 

The context package is very useful and the topic is very broad, it deserves its separate page. But for now it's good to keep these two points in mind regarding `defer` and `context`.

---

| [Previous](9-10-11.md) | [Home](../../index.md) | [Next](../more-types/1.md) |